PROTOCOL.md

```diff
@@ -0,0 +1,347 @@
+that all the details in this process are the same as previous models.
+Therefore, we will elide many details to make it easier to reason about. This
+does not imply these details are dropped under this model.
+
+```mermaid
+graph RL
+    subgraph parties
+        A(operator)
+        B(targetsrv)
+    end
+
+    subgraph host
+        C(enarx run)
+        D(enarx srv)
+        E(keep)
+
+	    style E fill:#0f0,stroke:#333,stroke-width:4px
+    end
+
+    A -. "1. enarx push" .-> B
+    B -. "2. pubₜ" .-> A
+    A -. "3. (uri, pubₜ)" .-> C
+
+    C -.-> D
+    D -.-> E
+
+    E -- "4. attest" --> B
+    B -- "5. manifest" --> E
+    E -. "6. canonicalize" .-> E
+    E -- "7. canon, [eitemhash]" --> B
+    B -- "8. [items]" --> E
+```
+
+1. Each party in the computation uses an agreed upon `enarx` build. Each party
+does an `enarx push` to store their `items` in their own `targetsrv`.
+2. Storing `items` in a `targetsrv` produces a public key (`pubₜ`) which
+references them.
+3. Each party submits its `target` configuration which are combined into
+`targets`. The `targets` input is supplied to the `enarx run` command, which
+can be executed either manually or by a scheduler. A `keep` is instantiated
+like normal.
+4. The `keep` attests to each `targetsrv`.
+5. Each `targetsrv` reports the manifest for all the items it is willing to
+deploy.
+6. The canonicalized manifest (`canon`) is produced by merging the `manifest`
+from each `targetsrv`.
+7. This step is where the magic happens. The `keep` requests the required
+`items` from each `targetsrv`. As part of this request, it provides the `canon`
+which circumscribes precisely how the final keep will be assembled, including
+all the `items` from all the `targets`. Each `targetsrv` can evaluate the final
+state of the `keep` before agreeing to supply its `items`.
+8. If the `targetsrv` agrees that the `items` can be used to compose the
+`canon`, the `items` will be released for use in the `keep`.
+
+## Model #4: Provable Function
+
+**TODO**
+
+# Discussion
+
+## Caching
+
+This protocol theoretically allows for the caching of `items` on the host side.
+The flow looks something like this:
+
+```mermaid
+graph TB
+    A(keep)
+    B(targetsrv)
+    C(host cache)
+
+    A -. "1. attest/manifest" .- B
+    A -. "2. check cache" .- C
+    A -. "3. get uncached items" .- B
+    A -. "4. insert encrypted(kₗ, item)" .- C
+```
+
+The cache would contain only encrypted items, so we wouldn't leak plaintext.
+Further, because we don't have `kₗ` and **must** fetch it from the `targetsrv`,
+the `targetsrv` still has access control over its items.
+
+However, a few downsides still remain.
+
+1. Tracking. It would be possible to tract which keeps are executed where.
+Further, if one could build a database of vulnerable `items` it would be
+possible to identify keeps that run them based on the items they look for in
+the cache.
+
+2. Offline Brute Force. Anyone who gets access to the cache can attempt to
```

Comment:
Add "The host is considered untrusted, and so any data available to it, including the cache, is considered untrusted.".